main/src/library/RedBlackSet.flix

```diff
@@ -0,0 +1,296 @@
+    ///
+    pub def foldLeft(f: (b, a) -> b & e, s: b, xs: RedBlackSet[a]): b & e =
+        let RedBlackSet(es) = xs;
+            RedBlackTree.foldLeft(f, s, es)
+
+    ///
+    /// Applies `f` to a start value `s` and all elements in `xs` going from right to left.
+    ///
+    /// That is, the result is of the form: `f(x1, ...f(xn-1, f(xn, s))...)`.
+    ///
+    pub def foldRight(f: (b, a) -> b & e, s: b, xs: RedBlackSet[a]): b & e =
+        let RedBlackSet(es) = xs;
+            RedBlackTree.foldRight(f, s, es)
+
+    ///
+    /// Alias for `reduceLeft`.
+    ///
+    pub def reduce(f: (a, a) -> a & e, xs: RedBlackSet[a]): Option[a] & e = reduceLeft(f, xs)
+
+    ///
+    /// Applies `f` to all elements in `xs` going from left to right until a single value `v` is obtained.  Returns `Some(v)`.
+    /// That is, the result is of the form: `Some(f(...f(f(x1, x2), x3)..., xn))`
+    /// Returns `None` if `xs` is the empty set.
+    ///
+    /// TODO: Fix, how to get first value as initial?
+    pub def reduceLeft(f: (a, a) -> a & e, xs: RedBlackSet[a]): Option[a] & e =
+        if (isEmpty(xs)) None
+        else Some(foldLeft(f, empty(), xs))
+
+    ///
+    /// Applies `f` to all elements in `xs` going from right to left until a single value `v` is obtained.  Returns `Some(v)`.
+    /// That is, the result is of the form: `Some(f(x1, ...f(xn-2, f(xn-1, xn))...))`
+    /// Returns `None` if `xs` is the empty set.
+    ///
+    pub def reduceRight(f: (a, a) -> a & e, xs: RedBlackSet[a]): Option[a] & e = ???
+
+    ///
+    /// Returns the number of elements in `xs` that satisfy the predicate `f`.
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def count(f: a -> Bool, xs: RedBlackSet[a]): Int32 =
+        fold((acc, x) -> if (f(x)) 1 + acc else acc, 0, xs)
+
+    ///
+    /// Returns the union of the elements in `xs`.
+    ///
+    pub def flatten(xs: RedBlackSet[RedBlackSet[a]]): RedBlackSet[a] = ???
+
+    ///
+    /// Returns `true` if and only if at least one element in `xs` satisfies the predicate `f`.
+    ///
+    /// Returns `false` if `xs` is the empty set.
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def exists(f: a -> Bool, xs: RedBlackSet[a]): Bool =
+        let result = foldLeft((acc, x) -> if (f(x)) insert(x, acc) else acc, empty(), xs);
+            !isEmpty(result)
+
+    ///
+    /// Returns `true` if and only if all elements in `xs` satisfy the predicate `f`.
+    ///
+    /// Returns `true` if `xs` is the empty set.
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def forall(f: a -> Bool, xs: RedBlackSet[a]): Bool =
+        let result = foldLeft((acc, x) -> if (f(x)) insert(x, acc) else acc, empty(), xs);
+            size(xs) == size(result)
+
+    ///
+    /// Returns the union of `xs` and `ys`.
+    ///
+    pub def union(xs: RedBlackSet[a], ys: RedBlackSet[a]): RedBlackSet[a] =
+        foldLeft((acc, x) -> insert(x, acc), ys, xs)
```

Comment:
Here we can be smart, as we talked about: If we know which tree is smaller, we can insert that into the larger.